src/components/VmDetails/cards/DisksCard/DiskImageEditor.js

```diff
@@ -0,0 +1,408 @@
+  constructor (props) {
+    super(props)
+    this.state = {
+      showModal: false,
+
+      id: undefined,
+      alias: props.suggestedName || '',
+      size: DISK_DEFAULTS.provisionedSize / 1024 ** 3,
+      storageDomain: props.suggestedStorageDomain || '',
+      format: DISK_DEFAULTS.format, // cow vs raw .. Allocation Policy (thin sparse vs preallocated !sparse)?
+    }
+    this.changesMade = false
+
+    this.open = this.open.bind(this)
+    this.close = this.close.bind(this)
+    this.composeNic = this.composeDisk.bind(this)
+    this.handleSave = this.handleSave.bind(this)
+
+    this.changeAlias = this.changeAlias.bind(this)
+    this.changeSize = this.changeSize.bind(this)
+    this.changeStorageDomain = this.changeStorageDomain.bind(this)
+    this.changeFormat = this.changeFormat.bind(this)
+  }
+
+  open (e) {
+    e.preventDefault()
+
+    const disk = this.props.disk // DiskAttachmentType
+    const diskInfo = disk
+      ? { // edit
+        id: disk.get('id'),
+        alias: disk.get('name'),
+        size: 0,
+        storageDomain: disk.get('storageDomainId'),
+        format: disk.get('format'),
+      }
+      : { // new
+        id: undefined,
+        alias: this.props.suggestedName,
+        size: DISK_DEFAULTS.provisionedSize / 1024 ** 3,
+        storageDomain: this.props.suggestedStorageDomain,
+        format: DISK_DEFAULTS.format,
+      }
+
+    this.setState({ showModal: true, ...diskInfo })
+    this.changesMade = false
+  }
+
+  close () {
+    this.setState({ showModal: false })
+  }
+
+  composeDisk () {
+    const { vm, disk } = this.props
+
+    const vmDiskInSameStorageDomain =
+      vm.get('disks') &&
+      vm.get('disks').find(disk => disk.get('storageDomainId') === this.state.storageDomain)
+
+    const iface =
+      (disk && disk.get('iface')) ||
+      (vmDiskInSameStorageDomain && vmDiskInSameStorageDomain.get('iface')) ||
+      'virtio_scsi'
+
+    const provisionedSize = disk
+      ? disk.get('provisionedSize') + this.state.size * 1024 ** 3
+      : this.state.size * 1024 ** 3
+
+    const newDisk = {
+      ...DISK_DEFAULTS,
+      attachmentId: disk && disk.get('attachmentId'),
+
+      iface,
+
+      id: this.state.id,
+      name: this.state.alias,
+      provisionedSize,
+      storageDomainId: this.state.storageDomain,
+      format: this.state.format,
+      sparse: this.state.format === 'cow',
+    }
+
+    if (disk && disk.get('type') !== 'image') {
+      newDisk.type = disk.get('type')
+      newDisk.provisionedSize = undefined
+    }
+
+    return newDisk
+  }
+
+  handleSave () {
+    if (this.changesMade) {
+      const newDisk = this.composeDisk()
+      this.props.onSave(this.props.vm.get('id'), newDisk)
+    }
+    this.close()
+  }
+
+  changeAlias ({ target: { value } }) {
+    this.setState({ alias: value })
+    this.changesMade = true
+  }
+
+  changeSize (e) {
+    if (e.target.value < 0) {
+      e.target.value = this.state.size
```

Comment:
It just prevents the user from setting a value less then 0.

Any ideas on how better to validate input?

Reply:
Yes, for example:

```JavaScript
  changeSize (e) {
        let value = parseInt(e.target.value);
        value = value < 0 || isNaN(value) ? 0 : value;
        this.setState({ size: value })
        this.changesMade = true
    }
```

Reply:
While that will avoid setting the component state's size below 0, it doesn't prevent the display from going below 0.  I'll probably need to switch to a controlled `FormControl` to do things in a more straightforward manner.

Reply:
Switched `FormControl` from `defaultValue` to `value` (making the field controlled) so that is only displayed what is held in the `DiskImageEditor`'s state.  Makes the `changeSize` function very simple.